demystifying_regular_expressions/demystifying_regular_expressions.md

```diff
@@ -0,0 +1,305 @@
+
+<div class = "care">
+<b style="color: rgb(var(--color-highlight));">A little encouragement...</b><br>
+
+Regular expressions require a bit of effort to learn, and even people who use regular expressions often will end up making mistakes and using online tools for help.  
+
+If you work with strings (character data) frequently, whether that's from free-text entries in a REDCap database, text from web-scraped blogs, file names from a huge dataset, or clinical notes from an electronic health record, the learning curve is definitely worth it!  
+
+</div>
+
+The goal of regular expressions is to describe patterns based in:
+
+* which kinds of characters occur, and 
+* if they are optional or mandatory,
+* how many of that kind of character to expect, and
+* in what order.  
+
+You can also put in a "capturing" element, to capture a specific part of the pattern (in our  example, extracting the patient identifier and/or the sample number).
+
+Some regular expressions are short and simple, and some can be many lines of text long.
+
+## Where Can You Use Regular Expressions?
+
+Regex is often used in **code**.  
+
+For example, you might write Python code that searches for patterns that look like medical record numbers and redacts them, replacing them with "9999999".  
+
+Or you could write R code that takes the file names from a directory full of video files with specially formatted file names like `subj412834_20230412_ALC.mp4` and makes a table of the information those file names contain (like subject id, date, initials of who made the interview recording, etc.).
+
+Regular expressions can also be used in **text editors**.  
+
+For example, this module was written using a text editor called [Visual Studio Code](https://code.visualstudio.com/) (or VS Code).  This text editor allows the user to search for patterns using regular expressions.  For example, the author can look for every time a word appears between single asterisks (which will mean the text will appear in italics in the module when a learner reads **).  
+
+VS Code also supports search and replace.  For example, if there's a style rule that indicates that words in *italics* (like `*regular*`) should also be in bold, the author can ask the text editor to find all the places where there's a word enclosed in single asterisks and enclose that word instead in triple asterisks (like `***regular***`, which will make the word render in ***bold italic***).
+
+There are also some popular **command line tools** that use regular expressions, like `sed`, `grep`, and `awk`.
+
+## An Example Regular Expression
```

Comment:
Structural suggestion: 
There is really one big example here, with 3 components:
1. breaking down the pattern of what "looks like" a website
2. parsing the code for that pattern in regex
3. showing that that pattern does match the things we want it to match and doesn't match things we don't.

I suggest reordering it this way, but if this makes less sense to you, please push back.

---

## An Example Regular Expression
Hey, let's investigate a regular expression for the pattern that makes up a url. We want it to match these [list] but not these [list].
### Describe the pattern in words
current contents of lines 130-146, describing in English what the pattern is
### There exists code that describes the same pattern in words
Introduce Regex 101. If you asked someone who knows regex to write down your pattern, this is what they would give you, and you can check that it works
### How does the code encode the pattern?
Content from current lines 178-207

Reply:
updated in commit!